### DISCUSSION TITLE
How to add middleware in different places?

### DISCUSSION THREAD

**Discussion by mht2953658596** [fastapi/fastapi#10142] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request, Depends
app = FastAPI()
app.add_middleware(
	CORSMiddleware,
	allow_origins=["*"],
	allow_credentials=False,
	allow_methods=["*"],
	allow_headers=["*"],
)


from fastapi import Request, Response
import time

from main import app


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
	start_time = time.time()
	response = await call_next(request)
	process_time = time.time() - start_time
	response.headers["X-Process-Time"] = str(process_time)
	return response
```

### Description

I've noticed that middleware functions in FastAPI need to be in the same .py file as the code containing app = FastAPI() in order for the middleware function to work.

in main.py:
```python
from fastapi import FastAPI, Request, Depends
app = FastAPI()
app.add_middleware(
	CORSMiddleware,
	allow_origins=["*"],
	allow_credentials=False,
	allow_methods=["*"],
	allow_headers=["*"],
)
````

This cross-origin middleware is effective.

However, I created a directory named "middleware" at the same level as the main.py, and inside that directory, there's a file named "demo_middleware.py".
The directory structure is as follows:
-main.py
-middleware
--__init__.py
--demo_middleware.py

in demo_middleware.py
```python
from fastapi import Request, Response
import time

from main import app


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
	start_time = time.time()
	response = await call_next(request)
	process_time = time.time() - start_time
	response.headers["X-Process-Time"] = str(process_time)
	return response

```

 However, this middleware function for measuring time isn't functioning.

### Operating System

Windows

### FastAPI Version

0.100.0

### Pydantic Version

1.10.7

### Python Version

Python 3.8.16

**Comment by vedil** [comment]
you can manually like you mentioned for CORS some like  `app.add_middleware(demo_middleware.Timer)`

you can follow https://github.com/tiangolo/fastapi/issues/397 to get more idea

**Comment by YuriiMotov** [comment, marked as answer]
Remove `from main import app` and `@app.middleware("http")` from `demo_middleware` module, then import your middleware to the `main` module and add middleware to the stack using
```
app.middleware("http")(demo_middleware.add_process_time_header)
```
or
```
app.add_middleware(BaseHTTPMiddleware, dispatch=demo_middleware.add_process_time_header)
```

Another way is to convert your middleware into the [pure ASGI middleware](https://www.starlette.io/middleware/#pure-asgi-middleware) and add it using `app.add_middleware(YourMiddlewareClass)`
